MACRO GDP reporting correction and documentation updated

GamzeUnlu95 opened the pull request:
This PR:

- Corrects the reporting of GDP by introducing the` GDP_corr `parameter which includes the deduction of `trade_costs` coming from MESSAGE. 

- Updates MESSAGEix based total_cost parameter in iteration with MACRO. 

- Extends the documentation of the capital constraint section in MACRO to explain the reasoning behind the changes.

behnam-zakeri commented on message_ix/model/MACRO/macro_core.gms:
```diff
@@ -160,11 +160,20 @@ SUM(node_active,
+* decrease of consumption and investment in the certain year as well as by the reduction in capacity to invest and grow the economy for later years. 
+* 
+* .. math:: Y_{n, y} = C_{n, y} + I_{n, y} + {EC}_{n, y} \qquad \forall{n, y}
+* .. math:: EC_{n, y} = COST NODAL_{n, y} + import cost_{r, y} - export cost_{n, y} \qquad \forall{n, y}
```
Two equations introduced here, i.e., EC and GDP are not followed by their respective mathematical formulation. EC comes a some lines later, while GDP is not calculated here.

GamzeUnlu95 replied:
Thanks this is a good point. Instead of EC, total_cost was meant actually and EC is approximated based on total_cost as in the below formula in Equation COST_ENERGY section of the documentation. So I changed this part of the formulation. 

GDP calculation is also explained in this section since it used to be the same as Y (income) but now there is a change in the calculation. The trade balance is considered in Y (income) and therefore not again in the GDP.

behnam-zakeri commented on message_ix/model/MESSAGE-MACRO_run.gms:
```diff
@@ -114,6 +114,7 @@ Parameters
     price_diff_rel(*,node,sector,year_all)
 
     report_iteration(iteration,*)
+    trade_cost_detail(node, commodity, year_all)              'net of commodity import costs and commodity export revenues by commodity, node and year'
```
Perhaps better to describe with the same order of indexes, i.e., "node, commodity, and year"

behnam-zakeri reviewed (commented):
Thanks @GamzeUnlu95 for drafting this correction. In my view, this is an enhancement; just a few points before merging:
- I suggested some minor edits in the GAMS files as inline comments
- More importantly, can you develop a test that show this correction. For example, a test that asserts that the calculated GDP should be equal to a certain value?
- I believe after applying these changes, some overnight tests may not pass because the objective value may change. I can't see that at the moment as the PR needs to be rebased and the tests should be run again.
Overall, nice contribution.

behnam-zakeri commented on message_ix/model/MESSAGE-MACRO_run.gms:
```diff
@@ -199,6 +202,10 @@ DISPLAY enestart, eneprice, total_cost ;
 *----------------------------------------------------------------------------------------------------------------------*
 
+* update total energy system costs by node and time with information from latest MESSAGE run
+total_cost(node_macro, year) = COST_NODAL_NET.L(node_macro, year) / 1000 ;
```
I see that here we introduce a new MESSAGE parameter (`total_cost`) to report an existing variable (`COST_NODAL_NET`) only with a unit conversion. As the units are not explicit in MESSAGE, it is not clear to me the use case of this new parameter? In general, I believe it's better not to introduce too many parameters in the output file if the same information can be seen in another parameter/variable.

GamzeUnlu95 replied:
`total_cost ` parameter is not introduced in this code but it is already introduced in `macro_data_load.gms` and used in macro_core.gms as follows:

```
COST_ENERGY(node_active, year) $ (NOT macro_base_period(year))..
EC(node_active, year) =E=
(total_cost(node_active, year)/1000
+ SUM(sector, eneprice(node_active, sector, year) * 1E-3 * (PHYSENE(node_active, sector, year) - enestart(node_active, sector, year)))
+ SUM(sector, eneprice(node_active, sector, year) * 1E-3 / enestart(node_active, sector, year) * (PHYSENE(node_active, sector, year) - enestart(node_active, sector, year)) * (PHYSENE(node_active, sector, year) - enestart(node_active, sector, year))))
;
```
Here the purpose of adding that line was not unit conversion but to update this parameter based on the COST_NODAL_NET in each iteration with MESSAGE.

behnam-zakeri replied:
Thanks for the explanation.

gidden commented:
Hi @GamzeUnlu95 - I understand there are still some requirements left for this to be merged. Would you say it is available to try to use on a trial basis?

gidden commented:
Hi folks, I wanted to report back after having tried to use this PR in some of my ongoing work.

Observed issue: Running a mitigation scenario with a calibrated scenario causes MESSAGE LP infeasibility within MESSAGE-MACRO iteration.

Suggestion: Before merging this, run a test case on the current ENGAGE set up where MACRO is recalibrated with these changes and confirm that mitigation cases can still converge.

More specifics:

It's hard for me to pin point exactly why this PR causes my failure, but I am combining a number of updates - I start from the new R15 model, apply 1-year time steps for early periods, and calibrate using this MACRO branch. I did not observe any issues while creating baselines or non-mitigation scenarios.

khaeru commented:
> MESSAGE LP infeasibility within MESSAGE-MACRO iteration

Can you please upload a snippet from the GAMS .lst file to indicate which variables are infeasible?

gidden commented:
That file has already been overwritten. I can try later to reproduce - can you tell me what to `grep` for?

gidden commented:
But this is a good opportunity to plug our need for better tooling when dealing with infeasibilities. It would be great to have some first (or even second) class support to say, dump out *why* a model was infeasible at the end of a run that is persistable.

khaeru commented:
> can you tell me what to `grep` for?

https://github.com/iiasa/message_ix/issues/418#issuecomment-720388779 for example

OFR-IIASA commented:
I have pulled these changes onto the latest main branch, updated the macro-calibration xlsx but then encountered errors.

```
**** Exec Error at line 544: log: FUNC SINGULAR: x = 0
**** Exec Error at line 544: overflow in * operation (mulop)
```

Has anyone encountered such issues? Was this PR already tested with the calibration? I have added the macro_run.lst file as a .txt file if this helps with debugging.
[MACRO_run.txt](https://github.com/iiasa/message_ix/files/9553986/MACRO_run.txt)

GamzeUnlu95 commented:
> I have pulled these changes onto the latest main branch, updated the macro-calibration xlsx but then encountered errors.
> 
> ```
> **** Exec Error at line 544: log: FUNC SINGULAR: x = 0
> **** Exec Error at line 544: overflow in * operation (mulop)
> ```
> 
> Has anyone encountered such issues? Was this PR already tested with the calibration? I have added the macro_run.lst file as a .txt file if this helps with debugging. [MACRO_run.txt](https://github.com/iiasa/message_ix/files/9553986/MACRO_run.txt)

Thanks for adding the .lst file. I attempted to test this PR with calibration a couple of months ago and at that time there was an issue with the macro-calibration xlsx in the main branch and I could not complete it. I need to find and allocate time soon to test this PR again and finalize it.

GamzeUnlu95 commented:
I tested this PR with the ENGAGE setup by using the model `ENGAGE_SSP2_v4.1.7_T4.5_r3.1_test`. I did not experience any issues with the calibration of the baseline. I ran the mitigation case with 600 budget and was able to get a solution for `EN_NPi2020_600`. However, in Step 3 of `EN_NPi2020_400`, in the third iteration, there is an execution error division by eps as in this [MESSAGE-MACRO_run.log](https://github.com/iiasa/message_ix/files/10959553/MESSAGE-MACRO_run.log) and [MESSAGE-MACRO_run_lst](https://github.com/iiasa/message_ix/files/10959566/MESSAGE-MACRO_run_lst.txt) file. Not sure if this is directly related to the changes that are made in this PR. The line indicated in the error message is the following: 
```
price_diff_rel(iteration,node_macro,sector,year)$( price_init(node_macro,sector,year) ) =
5455      price_diff_abs(iteration,node_macro,sector,year) / price_init(node_macro,sector,year)
```

glatterf42 commented:
At today's message meeting, this PR was suggested to me as a possible fix for #752. As far as I can tell, this PR does not enable usage of user-defined units. However, I rebased this onto current main so if you want to run any tests again to see if this can be merged, this might be a good time, @GamzeUnlu95, @OFR-IIASA, @gidden.

khaeru commented:
This PR may even predate the PR template, but it is missing one item that appears there: "Update release notes". I've added this to the PR description above.

These notes are especially important when the behaviour and outputs of the core GAMS code change, as in this PR. Users *must* have an opportunity to notice and understand how their model outputs may be affected by a new release.

glatterf42 commented:
As discussed in yesterday's weekly meeting, we hold off on merging these changes because we don't trust our current test suite to adequately verify that the changes won't break anything. As @gidden suggested (please correct me if I misunderstood), we should set up a testing workflow that runs a full emission baseline and a peak budget scenario. Since these runs might take a while, we would ideally keep them optional and either run them as part of nightly tests or enable them on PRs that affect the GAMS code. This would allow us to trust the changes here.

@Jihoon noted that the line currently causing the test failure reported by @GamzeUnlu95 had caused issues before, so was removed locally in his installation without any noticeable negative consequences. No one else in the room knew what this line might be good for, so in a separate PR, we could try to remove it -- and test that with the new test cases proposed above.

gidden commented:
Thanks @glatterf42. My suggestion is to follow @khaeru's example [here](https://github.com/iiasa/message_data/actions/runs/8746996958/workflow) and combine this with @OFR-IIASA's ENGAGE process script [here](https://github.com/iiasa/message_data/blob/ssp_dev/message_data/projects/engage/__init__.py). 

I would suggest to run 

1. a baseline scenario
2. a full century budget scenario
3. a peak-budget scenario

And compare a few key metrics on the output excel (e.g., co2 emissions, kyoto emissions, gdp, primary energy, final energy).

khaeru commented:
> My suggestion is to follow @khaeru's example [here](https://github.com/iiasa/message_data/actions/runs/8746996958/workflow) and combine this with @OFR-IIASA's ENGAGE process script [here](https://github.com/iiasa/message_data/blob/ssp_dev/message_data/projects/engage/__init__.py).

While I generally agree with this idea directionally and the basic observation underlying it, to implement these ideas would involve, conservatively, at least 2 person-weeks of work. Either we accept that this PR sits until someone is assigned/freed-up to spend that time, or we find a simpler, manual check (and someone to do it) that can assure that this PR will not change *[expected behaviour]*.

In order to avoid derailing this particular PR with that broader discussion, I'll open a new issue.
